src/DotNetWorker.Grpc/GrpcFunctionsHostLogger.cs

```diff
@@ -1,101 +1,143 @@
-
-            _channelWriter.TryWrite(response);
-        }
-
-        private static Level ToRpcLogLevel(LogLevel logLevel)
-        {
-            return logLevel switch
-            {
-                LogLevel.Trace => Level.Trace,
-                LogLevel.Debug => Level.Debug,
-                LogLevel.Information => Level.Information,
-                LogLevel.Warning => Level.Warning,
-                LogLevel.Error => Level.Error,
-                LogLevel.Critical => Level.Critical,
-                _ => Level.None,
-            };
-        }
-
-        private class EmptyDisposable : IDisposable
-        {
-            public static IDisposable Instance = new EmptyDisposable();
-
-            public void Dispose()
-            {
-            }
-        }
-    }
-}
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Channels;
+using Azure.Core.Serialization;
+using Microsoft.Azure.Functions.Worker.Diagnostics;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+using Microsoft.Azure.Functions.Worker.Logging.ApplicationInsights;
+using Microsoft.Azure.Functions.Worker.Rpc;
+using Microsoft.Extensions.Logging;
+using static Microsoft.Azure.Functions.Worker.Grpc.Messages.RpcLog.Types;
+
+namespace Microsoft.Azure.Functions.Worker.Logging
+{
+    /// <summary>
+    /// A logger that sends logs back to the Functions host.
+    /// </summary>
+    internal class GrpcFunctionsHostLogger : ILogger
+    {
+        private readonly string _category;
+        private readonly ChannelWriter<StreamingMessage> _channelWriter;
+        private readonly IExternalScopeProvider _scopeProvider;
+        private readonly ObjectSerializer _serializer;
+
+        public GrpcFunctionsHostLogger(string category, ChannelWriter<StreamingMessage> channelWriter, IExternalScopeProvider scopeProvider, ObjectSerializer serializer)
+        {
+            _category = category ?? throw new ArgumentNullException(nameof(category));
+            _channelWriter = channelWriter ?? throw new ArgumentNullException(nameof(channelWriter));
+            _scopeProvider = scopeProvider ?? throw new ArgumentNullException(nameof(scopeProvider));
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+        }
+
+        public IDisposable BeginScope<TState>(TState state)
+        {
+            // The built-in DI wire-up guarantees that scope provider will be set.
+            return _scopeProvider!.Push(state);
+        }
+
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return logLevel != LogLevel.None;
+        }
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            if (eventId == LogConstants.MetricEventId)

```

Comment:
Can we check the event *name* here instead of the *id*? Then we can remove that `-1` change in the test. Otherwise, I worry this is a breaking change and that anyone currently using an event with id of `1` will now go down this path.